File: test/samples/vulnerable-app.js
/**
 * Sample Vulnerable JavaScript Application
 * This file contains multiple cryptographic vulnerabilities for testing
 */

const crypto = require('crypto');
const fs = require('fs');

// CRITICAL: MD5 Hash - Cryptographically Broken
function hashPasswordMD5(password) {
    const hash = crypto.createHash('md5');
    hash.update(password);
    return hash.digest('hex');
}

// CRITICAL: SHA-1 Hash - Cryptographically Broken
function generateTokenSHA1(data) {
    const hash = crypto.createHash('sha1');
    hash.update(data);
    return hash.digest('hex');
}

// HIGH: RSA 1024-bit - Quantum Vulnerable
function generateRSA1024KeyPair() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
        modulusLength: 1024,
        publicKeyEncoding: {
            type: 'spki',
            format: 'pem'
        },
        privateKeyEncoding: {
            type: 'pkcs8',
            format: 'pem'
        }
    });
    return { publicKey, privateKey };
}

// HIGH: RSA 2048-bit - Quantum Vulnerable (better but still vulnerable)
function generateRSA2048KeyPair() {
    return crypto.generateKeyPairSync('rsa', {
        modulusLength: 2048,
        publicKeyEncoding: { type: 'spki', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
    });
}

// HIGH: ECDSA - Quantum Vulnerable
function signWithECDSA(data, privateKey) {
    const sign = crypto.createSign('ecdsa-with-SHA256');
    sign.update(data);
    return sign.sign(privateKey, 'hex');
}

// MEDIUM: 3DES - Deprecated
function encrypt3DES(text, key, iv) {
    const cipher = crypto.createCipheriv('des-ede3', key, iv);
    let encrypted = cipher.update(text, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    return encrypted;
}

// MEDIUM: DES - Deprecated and Weak
function encryptDES(text, key, iv) {
    const cipher = crypto.createCipheriv('des', key, iv);
    let encrypted = cipher.update(text, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    return encrypted;
}

// MEDIUM: DSA - Quantum Vulnerable
function generateDSAKeyPair() {
    return crypto.generateKeyPairSync('dsa', {
        modulusLength: 2048,
        divisorLength: 256,
        publicKeyEncoding: { type: 'spki', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
    });
}

// HIGH: Diffie-Hellman - Quantum Vulnerable
function createDHKeyExchange() {
    const dh = crypto.createDiffieHellman(2048);
    dh.generateKeys();
    return {
        publicKey: dh.getPublicKey('hex'),
        privateKey: dh.getPrivateKey('hex')
    };
}

// Application usage examples
class VulnerableAuthSystem {
    constructor() {
        this.users = new Map();
    }

    // CRITICAL: Storing password hashes with MD5
    registerUser(username, password) {
        const passwordHash = hashPasswordMD5(password);
        this.users.set(username, passwordHash);
        console.log(`User ${username} registered with MD5 hash`);
    }

    // CRITICAL: Using SHA-1 for session tokens
    createSession(username) {
        const timestamp = Date.now().toString();
        const sessionToken = generateTokenSHA1(username + timestamp);
        console.log(`Session token created with SHA-1`);
        return sessionToken;
    }

    // HIGH: Using RSA 1024 for encryption
    encryptData(data) {
        const { publicKey } = generateRSA1024KeyPair();
        const encrypted = crypto.publicEncrypt(publicKey, Buffer.from(data));
        return encrypted.toString('base64');
    }
}

// Example API endpoints (Express-style)
const routes = {
    // CRITICAL: MD5 for file integrity
    '/upload': function(req, res) {
        const fileHash = crypto.createHash('md5')
            .update(req.body.file)
            .digest('hex');
        res.json({ checksum: fileHash });
    },

    // HIGH: RSA 2048 for key exchange
    '/key-exchange': function(req, res) {
        const keys = generateRSA2048KeyPair();
        res.json({ publicKey: keys.publicKey });
    },

    // MEDIUM: 3DES encryption
    '/encrypt': function(req, res) {
        const key = Buffer.alloc(24);
        const iv = Buffer.alloc(8);
        const encrypted = encrypt3DES(req.body.data, key, iv);
        res.json({ encrypted });
    }
};

// Initialize vulnerable system
const auth = new VulnerableAuthSystem();
auth.registerUser('admin', 'password123');
const session = auth.createSession('admin');

// Export for testing
module.exports = {
    hashPasswordMD5,
    generateTokenSHA1,
    generateRSA1024KeyPair,
    generateRSA2048KeyPair,
    signWithECDSA,
    encrypt3DES,
    encryptDES,
    generateDSAKeyPair,
    createDHKeyExchange,
    VulnerableAuthSystem
};

File: test/samples/vulnerable-app.py
"""
Sample Vulnerable Python Application
This file contains multiple cryptographic vulnerabilities for testing
"""

import hashlib
import hmac
from cryptography.hazmat.primitives.asymmetric import rsa, dsa, ec
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
import os


# CRITICAL: MD5 Hash - Cryptographically Broken
def hash_password_md5(password):
    """Hash password using MD5 (VULNERABLE!)"""
    return hashlib.md5(password.encode()).hexdigest()


# CRITICAL: SHA-1 Hash - Cryptographically Broken
def generate_token_sha1(data):
    """Generate token using SHA-1 (VULNERABLE!)"""
    return hashlib.sha1(data.encode()).hexdigest()


# HIGH: RSA 1024-bit - Quantum Vulnerable
def generate_rsa_1024_keypair():
    """Generate RSA 1024-bit key pair (VULNERABLE!)"""
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=1024,
        backend=default_backend()
    )
    public_key = private_key.public_key()
    return private_key, public_key


# HIGH: RSA 2048-bit - Quantum Vulnerable
def generate_rsa_2048_keypair():
    """Generate RSA 2048-bit key pair (VULNERABLE to quantum)"""
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
        backend=default_backend()
    )
    public_key = private_key.public_key()
    return private_key, public_key


# HIGH: ECDSA - Quantum Vulnerable
def generate_ecdsa_keypair():
    """Generate ECDSA key pair (VULNERABLE to quantum)"""
    private_key = ec.generate_private_key(
        ec.SECP256R1(),
        default_backend()
    )
    public_key = private_key.public_key()
    return private_key, public_key


# MEDIUM: 3DES - Deprecated
def encrypt_3des(plaintext, key):
    """Encrypt using 3DES (DEPRECATED!)"""
    iv = os.urandom(8)
    cipher = Cipher(
        algorithms.TripleDES(key),
        modes.CBC(iv),
        backend=default_backend()
    )
    encryptor = cipher.encryptor()

    # Padding
    padding_length = 8 - (len(plaintext) % 8)
    padded_plaintext = plaintext + (chr(padding_length) * padding_length).encode()

    ciphertext = encryptor.update(padded_plaintext) + encryptor.finalize()
    return iv + ciphertext


# MEDIUM: DES - Deprecated and Weak
def encrypt_des(plaintext, key):
    """Encrypt using DES (DEPRECATED and WEAK!)"""
    # Note: cryptography library doesn't support DES directly
    # This is a placeholder showing the vulnerable pattern
    pass


# MEDIUM: DSA - Quantum Vulnerable
def generate_dsa_keypair():
    """Generate DSA key pair (VULNERABLE to quantum)"""
    private_key = dsa.generate_private_key(
        key_size=2048,
        backend=default_backend()
    )
    public_key = private_key.public_key()
    return private_key, public_key


# CRITICAL: HMAC with MD5
def hmac_md5(key, message):
    """Generate HMAC using MD5 (VULNERABLE!)"""
    return hmac.new(key.encode(), message.encode(), hashlib.md5).hexdigest()


# CRITICAL: HMAC with SHA-1
def hmac_sha1(key, message):
    """Generate HMAC using SHA-1 (VULNERABLE!)"""
    return hmac.new(key.encode(), message.encode(), hashlib.sha1).hexdigest()


class VulnerableAuthSystem:
    """Example vulnerable authentication system"""

    def __init__(self):
        self.users = {}

    # CRITICAL: Storing password hashes with MD5
    def register_user(self, username, password):
        """Register user with MD5 password hash (VULNERABLE!)"""
        password_hash = hash_password_md5(password)
        self.users[username] = password_hash
        print(f"User {username} registered with MD5 hash")

    # CRITICAL: Using SHA-1 for session tokens
    def create_session(self, username):
        """Create session token using SHA-1 (VULNERABLE!)"""
        import time
        timestamp = str(time.time())
        session_token = generate_token_sha1(username + timestamp)
        print(f"Session token created with SHA-1")
        return session_token

    # HIGH: Using RSA 1024 for encryption
    def encrypt_data(self, data):
        """Encrypt data using RSA 1024 (VULNERABLE!)"""
        private_key, public_key = generate_rsa_1024_keypair()
        from cryptography.hazmat.primitives.asymmetric import padding

        encrypted = public_key.encrypt(
            data.encode(),
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        return encrypted


class VulnerableAPIEndpoints:
    """Example API endpoints with crypto vulnerabilities"""

    # CRITICAL: MD5 for file integrity
    def upload_file(self, file_content):
        """Upload file with MD5 checksum (VULNERABLE!)"""
        checksum = hashlib.md5(file_content).hexdigest()
        return {"checksum": checksum}

    # HIGH: RSA 2048 for key exchange
    def key_exchange(self):
        """Key exchange using RSA 2048 (VULNERABLE to quantum)"""
        private_key, public_key = generate_rsa_2048_keypair()

        pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        return {"publicKey": pem.decode()}

    # MEDIUM: 3DES encryption
    def encrypt_message(self, message):
        """Encrypt message using 3DES (DEPRECATED!)"""
        key = os.urandom(24)  # 192-bit key for 3DES
        encrypted = encrypt_3des(message.encode(), key)
        return {"encrypted": encrypted.hex()}


# Usage examples
if __name__ == "__main__":
    # Initialize vulnerable system
    auth = VulnerableAuthSystem()
    auth.register_user("admin", "password123")
    session = auth.create_session("admin")

    # API endpoints
    api = VulnerableAPIEndpoints()

    # File upload with MD5
    result = api.upload_file(b"test file content")
    print(f"File checksum: {result['checksum']}")

    # Key exchange with RSA 2048
    keys = api.key_exchange()
    print("Public key generated")

    # Encrypt with 3DES
    encrypted = api.encrypt_message("sensitive data")
    print(f"Encrypted message: {encrypted['encrypted'][:50]}...")
